x-pack/plugins/security/server/lib/authentication/providers/oidc.ts

```diff
@@ -0,0 +1,544 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { Legacy } from 'kibana';
+import { canRedirectRequest } from '../../can_redirect_request';
+import { getErrorStatusCode } from '../../errors';
+import { AuthenticationResult } from '../authentication_result';
+import { DeauthenticationResult } from '../deauthentication_result';
+import { AuthenticationProviderOptions, BaseAuthenticationProvider } from './base';
+
+/**
+ * The state supported by the provider (for the OpenID Connect handshake or established session).
+ */
+interface ProviderState {
+  /**
+   * Unique identifier of the OpenID Connect request initiated the handshake used to mitigate
+   * replay attacks.
+   */
+  nonce?: string;
+
+  /**
+   * Unique identifier of the OpenID Connect request initiated the handshake used to mitigate
+   * CSRF.
+   */
+  state?: string;
+
+  /**
+   * URL to redirect user to after successful OpenID Connect handshake.
+   */
+  nextURL?: string;
+
+  /**
+   * Elasticsearch access token issued as the result of successful OpenID Connect handshake and that should be provided
+   * with every request to Elasticsearch on behalf of the authenticated user. This token will eventually expire.
+   */
+  accessToken?: string;
+
+  /**
+   * Once the elasticsearch access token expires the refresh token is used to get a new pair of access/refresh tokens
+   * without any user involvement. If not used this token will eventually expire as well.
+   */
+  refreshToken?: string;
+}
+
+/**
+ * Defines the shape of an incoming OpenID Connect Request
+ */
+type OIDCIncomingRequest = Legacy.Request & {
+  payload: { iss?: string };
+  query: {
+    iss?: string;
+    code?: string;
+    state?: string;
+    error?: string;
```

Comment:
Yes, these are passed in ES to be handled there in order to construct the error message that will be shown ( after state is checked ). I added these because I used 
```
request is OIDCIncomingRequest 
```
in `function isOIDCIncomingRequest(request: Legacy.Request)`  but I can remove them if you think it's redundant